### PR TITLE
Add serde support to date time

### PR DESCRIPTION
**Pull request by thomas-k-cameron**
## Motivation and Context
This is a child PR of https://github.com/awslabs/smithy-rs/pull/2616

The changes that this PR introduces is same as the ones that were merged to `unstable-serde-support` branch before.

Initially, we tried to make commit to unstable-serde-support branch and merge changes one by one in small PRs. However, in order to make it up to date with the main branch, we would need to go through a large PR of over 700 files.

Thus, I decided to create individual PRs that commits directly to `main` branch.

## Description
- Implements `serde` support to `DateTime`

## Testing
- Test checks whether the serialized/de-serialized data matches with the expected value

## Checklist
NA

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: rust-runtime/aws-smithy-types/src/date_time/mod.rs
Comment (jdisanti) [inline rust-runtime/aws-smithy-types/src/date_time/mod.rs:637]: Should there also be a Cargo.toml change that adds the ciborium crate in this PR?

File: rust-runtime/aws-smithy-types/src/date_time/mod.rs
Comment (jdisanti) [inline rust-runtime/aws-smithy-types/src/date_time/mod.rs:612]: Consider moving these to the corresponding ser/de modules so that the cfg flag isn't needed on them.

File: rust-runtime/aws-smithy-types/src/date_time/de.rs
Comment (jdisanti) [inline rust-runtime/aws-smithy-types/src/date_time/de.rs:15]: It looks like this unexpected state is never set, so it's probably not needed.

**Review by jdisanti — COMMENTED** [review]
Looks good. Just some minor things.

Could you also add a `cargo` invocation to this script that runs tests with `aws_sdk_unstable` and the two serde features enabled? https://github.com/awslabs/smithy-rs/blob/cbbbec4a4892e301ebde3af5cc852aa48d82e608/rust-runtime/aws-smithy-types/additional-ci#L7

File: rust-runtime/aws-smithy-types/src/date_time/mod.rs
Comment (thomas-k-cameron) [reply rust-runtime/aws-smithy-types/src/date_time/mod.rs:637]: I think ciborium got lost somewhere when I was separating it into sub-PR. Let me fix.

File: rust-runtime/aws-smithy-types/src/date_time/mod.rs
Comment (thomas-k-cameron) [reply rust-runtime/aws-smithy-types/src/date_time/mod.rs:612]: ok

File: rust-runtime/aws-smithy-types/src/date_time/de.rs
Comment (thomas-k-cameron) [reply rust-runtime/aws-smithy-types/src/date_time/de.rs:15]: It was meant to protect from something un-expected(it was set after subsec nano), but I think we don't need it.

File: rust-runtime/aws-smithy-types/src/date_time/mod.rs
Comment (thomas-k-cameron) [reply rust-runtime/aws-smithy-types/src/date_time/mod.rs:613]: Let me fix it up.

Comment (thomas-k-cameron) [reply rust-runtime/aws-smithy-types/src/date_time/mod.rs]: fixed.

**Comment by thomas-k-cameron** [comment]
I don't know where the error is coming from.
```
    Checking serde v1.0.0
error[E0432]: unresolved import `core::nonzero`
   --> /opt/cargo/registry/src/github.com-1ecc6299db9ec823/serde-1.0.0/src/lib.rs:195:19
    |
195 |     pub use core::nonzero::{NonZero, Zeroable};
    |                   ^^^^^^^ could not find `nonzero` in `core`

error[E0432]: unresolved import `core::num::Zero`
   --> /opt/cargo/registry/src/github.com-1ecc6299db9ec823/serde-1.0.0/src/lib.rs:198:13
    |
198 |     pub use core::num::Zero;
    |             ^^^^^^^^^^^^^^^ no `Zero` in `num`
    |
help: consider importing one of these items instead
    |
198 |     pub use core::num::FpCategory::Zero;
    |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
198 |     pub use core::num::IntErrorKind::Zero;
    |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
198 |     pub use core::num::flt2dec::FullDecoded::Zero;
    |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
198 |     pub use core::num::fmt::Part::Zero;
    |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~

error[E0635]: unknown feature `zero_one`
  --> /opt/cargo/registry/src/github.com-1ecc6299db9ec823/serde-1.0.0/src/lib.rs:91:68
   |
91 | #![cfg_attr(feature = "unstable", feature(nonzero, specialization, zero_one))]
   |                                                                    ^^^^^^^^

Some errors have detailed explanations: E0432, E0635.
For more information about an error, try `rustc --explain E0432`.
error: could not compile `serde` due to 3 previous errors
warning: build failed, waiting for other jobs to finish...
error: process didn't exit successfully: `/opt/rustup/toolchains/nightly-2022-11-16-x86_64-unknown-linux-gnu/bin/cargo check --all-features --manifest-path aws-inlineable/Cargo.toml` (exit status: 101)
error: process didn't exit successfully: `/opt/rustup/toolchains/nightly-2022-11-16-x86_64-unknown-linux-gnu/bin/cargo hack check --all-features` (exit status: 1) 
```

The nightly compiler that you are using seems to be pretty old, maybe an update?

**Comment by thomas-k-cameron** [comment]
Oh, it's working now!

**Comment by thomas-k-cameron** [comment]
OK, it's working now.

File: rust-runtime/aws-smithy-types/src/date_time/de.rs
Comment (thomas-k-cameron) [reply rust-runtime/aws-smithy-types/src/date_time/de.rs:63]: Thanks for letting me know! 

I thought it is already covered by the existing tests.

Comment (thomas-k-cameron) [reply rust-runtime/aws-smithy-types/src/date_time/de.rs]: I figured out why it wasn't showing up any errors.

This branch doesn't have `[features]` key on `Cargo.toml`.
It seems like cargo just ignored anything that was feature-gated behind `serde-serialize` and `serde-deserialize`

**Comment by thomas-k-cameron** [comment]
I ran into few problem with non-human readable format.

- I realized that each data format has it's own way of implementing number type. You might lose some bits of information when you try to de-serialize it.
e.g. 
CBOR uses 128 bit integer to store number types.

- implementing `deser` for array that produces different data types is pretty complicated

I suggest we change `deser/ser` for non human readable format same as the human readable format. i.e. serialize them as RFC-3339 string and expect it when de-serializing.  
FYI, This is the same as the `chrono` crate.

**Comment by jdisanti** [comment]
What was the problem? Does it not call the different visit methods? We could just make them both `i64` if that's the case, and it would still be 5 bytes less than the string version.

**Comment by thomas-k-cameron** [comment]
You are right! 

It's working now. I don't know what I was thinking.

**Comment by thomas-k-cameron** [comment]
No idea where linker error is coming from.
```
 Checking examples/testing...
+ pushd testing
+ cargo check
    Updating crates.io index
   Compiling proc-macro2 v1.0.59
   Compiling quote v1.0.28
   Compiling unicode-ident v1.0.9
   Compiling autocfg v1.1.0
   Compiling libc v0.2.144
   Compiling syn v2.0.18
    Checking cfg-if v1.0.0
   Compiling version_check v0.9.4
    Checking once_cell v1.17.2
    Checking bytes v1.4.0
    Checking pin-project-lite v0.2.9
   Compiling log v0.4.18
   Compiling lock_api v0.4.9
    Checking itoa v1.0.6
   Compiling parking_lot_core v0.9.7
    Checking smallvec v1.10.0
   Compiling futures-core v0.3.28
    Checking scopeguard v1.1.0
   Compiling tokio v1.28.2
    Checking parking_lot v0.12.1
    Checking mio v0.8.8
    Checking socket2 v0.4.9
    Checking signal-hook-registry v1.4.1
    Checking num_cpus v1.15.0
    Checking tracing-core v0.1.31
   Compiling futures-task v0.3.28
   Compiling num-traits v0.2.15
   Compiling futures-util v0.3.28
   Compiling tokio-macros v2.1.0
   Compiling tracing-attributes v0.1.24
    Checking ryu v1.0.13
    Checking tracing v0.1.37
   Compiling num-integer v0.1.45
    Checking pin-utils v0.1.0
    Checking fnv v1.0.7
    Checking time-core v0.1.1
    Checking outref v0.5.1
    Checking vsimd v0.8.0
    Checking time v0.3.21
    Checking base64-simd v0.8.0
    Checking http v0.2.9
   Compiling slab v0.4.8
   Compiling indexmap v1.9.3
   Compiling cc v1.0.79
    Checking aws-smithy-types v0.0.0-smithy-rs-head (/home/build/workspace/aws-sdk/sdk/aws-smithy-types)
   Compiling semver v1.0.17
    Checking futures-sink v0.3.28
   Compiling futures-channel v0.3.28
   Compiling typenum v1.16.0
   Compiling ring v0.16.20
    Checking hashbrown v0.12.3
   Compiling httparse v1.8.0
    Checking tokio-util v0.7.8
   Compiling generic-array v0.14.7
    Checking try-lock v0.2.4
   Compiling memchr v2.5.0
    Checking tower-service v0.3.2
   Compiling crc32fast v1.3.2
    Checking want v0.3.0
    Checking h2 v0.3.19
    Checking http-body v0.4.5
   Compiling proc-macro-error-attr v1.0.4
    Checking httpdate v1.0.2
    Checking spin v0.5.2
   Compiling syn v1.0.109
    Checking percent-encoding v2.2.0
    Checking untrusted v0.7.1
    Checking hyper v0.14.26
   Compiling rustc_version v0.4.0
   Compiling proc-macro-error v1.0.4
   Compiling serde v1.0.163
    Checking either v1.8.1
    Checking bytes-utils v0.1.3
    Checking aho-corasick v1.0.1
    Checking aws-smithy-eventstream v0.0.0-smithy-rs-head (/home/build/workspace/aws-sdk/sdk/aws-smithy-eventstream)
   Compiling serde_derive v1.0.163
    Checking regex-syntax v0.7.2
   Compiling rustls v0.20.8
   Compiling serde_json v1.0.96
    Checking regex v1.8.3
    Checking aws-smithy-http v0.0.0-smithy-rs-head (/home/build/workspace/aws-sdk/sdk/aws-smithy-http)
    Checking block-buffer v0.10.4
    Checking crypto-common v0.1.6
    Checking sct v0.7.0
    Checking webpki v0.22.0
   Compiling pin-project-internal v1.1.0
    Checking base64 v0.21.2
    Checking subtle v2.5.0
   Compiling thiserror v1.0.40
    Checking xmlparser v0.13.5
    Checking lazy_static v1.4.0
    Checking digest v0.10.7
    Checking rustls-pemfile v1.0.2
    Checking pin-project v1.1.0
   Compiling extend v0.1.2
    Checking tokio-stream v0.1.14
   Compiling thiserror-impl v1.0.40
    Checking yansi v0.5.1
    Checking tower-layer v0.3.2
    Checking openssl-probe v0.1.5
    Checking diff v0.1.13
    Checking pretty_assertions v1.3.0
    Checking rustls-native-certs v0.6.2
    Checking tower v0.4.13
    Checking aws-smithy-async v0.0.0-smithy-rs-head (/home/build/workspace/aws-sdk/sdk/aws-smithy-async)
    Checking tokio-rustls v0.23.4
    Checking assert-json-diff v1.1.0
    Checking roxmltree v0.14.1
    Checking fastrand v1.9.0
    Checking hyper-rustls v0.23.2
    Checking aws-smithy-http-tower v0.0.0-smithy-rs-head (/home/build/workspace/aws-sdk/sdk/aws-smithy-http-tower)
   Compiling aws-types v0.0.0-smithy-rs-head (/home/build/workspace/aws-sdk/sdk/aws-types)
    Checking cpufeatures v0.2.7
    Checking zeroize v1.6.0
    Checking aws-credential-types v0.0.0-smithy-rs-head (/home/build/workspace/aws-sdk/sdk/aws-credential-types)
    Checking sha2 v0.10.6
    Checking aws-smithy-protocol-test v0.0.0-smithy-rs-head (/home/build/workspace/aws-sdk/sdk/aws-smithy-protocol-test)
    Checking form_urlencoded v1.1.0
   Compiling io-lifetimes v1.0.11
    Checking aws-smithy-client v0.0.0-smithy-rs-head (/home/build/workspace/aws-sdk/sdk/aws-smithy-client)
    Checking hex v0.4.3
    Checking hmac v0.12.1
   Compiling rustix v0.37.19
    Checking aws-sigv4 v0.0.0-smithy-rs-head (/home/build/workspace/aws-sdk/sdk/aws-sigv4)
    Checking tinyvec_macros v0.1.1
    Checking bitflags v1.3.2
    Checking linux-raw-sys v0.3.8
    Checking tinyvec v1.6.0
    Checking aws-sig-auth v0.0.0-smithy-rs-head (/home/build/workspace/aws-sdk/sdk/aws-sig-auth)
    Checking aws-endpoint v0.0.0-smithy-rs-head (/home/build/workspace/aws-sdk/sdk/aws-endpoint)
    Checking aws-http v0.0.0-smithy-rs-head (/home/build/workspace/aws-sdk/sdk/aws-http)
   Compiling crc32c v0.6.3
    Checking aws-smithy-json v0.0.0-smithy-rs-head (/home/build/workspace/aws-sdk/sdk/aws-smithy-json)
    Checking utf8parse v0.2.1
    Checking is-terminal v0.4.7
    Checking anstyle-parse v0.2.0
    Checking unicode-normalization v0.1.22
    Checking aws-smithy-xml v0.0.0-smithy-rs-head (/home/build/workspace/aws-sdk/sdk/aws-smithy-xml)
    Checking colorchoice v1.0.0
    Checking anstyle v1.0.0
    Checking urlencoding v2.1.2
    Checking anstyle-query v1.0.0
    Checking regex-syntax v0.6.29
    Checking unicode-bidi v0.3.13
    Checking idna v0.3.0
    Checking anstream v0.3.2
    Checking aws-smithy-query v0.0.0-smithy-rs-head (/home/build/workspace/aws-sdk/sdk/aws-smithy-query)
    Checking sha1 v0.10.5
    Checking md-5 v0.10.5
   Compiling heck v0.4.1
    Checking regex-automata v0.1.10
    Checking strsim v0.10.0
    Checking overload v0.1.1
    Checking clap_lex v0.5.0
   Compiling async-trait v0.1.68
    Checking matchers v0.1.0
    Checking clap_builder v4.3.1
    Checking nu-ansi-term v0.46.0
   Compiling clap_derive v4.3.1
    Checking aws-smithy-checksums v0.0.0-smithy-rs-head (/home/build/workspace/aws-sdk/sdk/aws-smithy-checksums)
    Checking aws-sdk-sts v0.29.0 (/home/build/workspace/aws-sdk/sdk/sts)
    Checking url v2.3.1
    Checking aws-sdk-sso v0.29.0 (/home/build/workspace/aws-sdk/sdk/sso)
    Checking tracing-log v0.1.3
    Checking sharded-slab v0.1.4
    Checking thread_local v1.1.7
    Checking aws-config v0.0.0-smithy-rs-head (/home/build/workspace/aws-sdk/sdk/aws-config)
    Checking tracing-subscriber v0.3.17
    Checking clap v4.3.1
    Checking aws-sdk-s3 v0.29.0 (/home/build/workspace/aws-sdk/sdk/s3)
    Checking testing-examples v0.1.0 (/home/build/workspace/aws-sdk/examples/testing)
    Finished dev [unoptimized + debuginfo] target(s) in 1m 21s
+ popd
+ for example in '*'
+ echo -e '\033[1;33mChecking examples/test-utils...\033[0m'
+ pushd test-utils
Checking examples/test-utils...
+ cargo check
    Updating crates.io index
   Compiling autocfg v1.1.0
   Compiling proc-macro2 v1.0.59
   Compiling unicode-ident v1.0.9
   Compiling quote v1.0.28
   Compiling libc v0.2.144
   Compiling syn v2.0.18
   Compiling version_check v0.9.4
    Checking cfg-if v1.0.0
    Checking bytes v1.4.0
   Compiling lock_api v0.4.9
   Compiling parking_lot_core v0.9.7
    Checking pin-project-lite v0.2.9
    Checking once_cell v1.17.2
    Checking scopeguard v1.1.0
   Compiling futures-core v0.3.28
    Checking smallvec v1.10.0
    Checking itoa v1.0.6
   Compiling log v0.4.18
   Compiling tokio v1.28.2
    Checking parking_lot v0.12.1
    Checking mio v0.8.8
    Checking socket2 v0.4.9
    Checking signal-hook-registry v1.4.1
    Checking num_cpus v1.15.0
   Compiling futures-task v0.3.28
   Compiling futures-util v0.3.28
    Checking tracing-core v0.1.31
   Compiling num-traits v0.2.15
    Checking ryu v1.0.13
    Checking pin-utils v0.1.0
   Compiling tokio-macros v2.1.0
   Compiling tracing-attributes v0.1.24
error: linking with `cc` failed: exit status: 1
  |
  = note: "cc" "-Wl,--version-script=/tmp/rustcabi7DK/list" "-m64" "/tmp/rustcabi7DK/symbols.o" "/home/build/workspace/aws-sdk/examples/test-utils/target/debug/deps/tracing_attributes-9b799731f627d96b.tracing_attributes.62e245ff-cgu.0.rcgu.o" "/home/build/workspace/aws-sdk/examples/test-utils/target/debug/deps/tracing_attributes-9b799731f627d96b.tracing_attributes.62e245ff-cgu.1.rcgu.o" "/home/build/workspace/aws-sdk/examples/test-utils/target/debug/deps/tracing_attributes-9b799731f627d96b.tracing_attributes.62e245ff-cgu.10.rcgu.o" "/home/build/workspace/aws-sdk/examples/test-utils/target/debug/deps/tracing_attributes-9b799731f627d96b.tracing_attributes.62e245ff-cgu.11.rcgu.o" "/home/build/workspace/aws-sdk/examples/test-utils/target/debug/deps/tracing_attributes-9b799731f627d96b.tracing_attributes.62e245ff-cgu.12.rcgu.o" "/home/build/workspace/aws-sdk/examples/test-utils/target/debug/deps/tracing_attributes-9b799731f627d96b.tracing_attributes.62e245ff-cgu.13.rcgu.o" "/home/build/workspace/aws-sdk/examples/test-utils/target/debug/deps/tracing_attributes-9b799731f627d96b.tracing_attributes.62e245ff-cgu.14.rcgu.o" "/home/build/workspace/aws-sdk/examples/test-utils/target/debug/deps/tracing_attributes-9b799731f627d96b.tracing_attributes.62e245ff-cgu.15.rcgu.o" "/home/build/workspace/aws-sdk/examples/test-utils/target/debug/deps/tracing_attributes-9b799731f627d96b.tracing_attributes.62e245ff-cgu.2.rcgu.o" "/home/build/workspace/aws-sdk/examples/test-utils/target/debug/deps/tracing_attributes-9b799731f627d96b.tracing_attributes.62e245ff-cgu.3.rcgu.o" "/home/build/workspace/aws-sdk/examples/test-utils/target/debug/deps/tracing_attributes-9b799731f627d96b.tracing_attributes.62e245ff-cgu.4.rcgu.o" "/home/build/workspace/aws-sdk/examples/test-utils/target/debug/deps/tracing_attributes-9b799731f627d96b.tracing_attributes.62e245ff-cgu.5.rcgu.o" "/home/build/workspace/aws-sdk/examples/test-utils/target/debug/deps/tracing_attributes-9b799731f627d96b.tracing_attributes.62e245ff-cgu.6.rcgu.o" "/home/build/workspace/aws-sdk/examples/test-utils/target/debug/deps/tracing_attributes-9b799731f627d96b.tracing_attributes.62e245ff-cgu.7.rcgu.o" "/home/build/workspace/aws-sdk/examples/test-utils/target/debug/deps/tracing_attributes-9b799731f627d96b.tracing_attributes.62e245ff-cgu.8.rcgu.o" "/home/build/workspace/aws-sdk/examples/test-utils/target/debug/deps/tracing_attributes-9b799731f627d96b.tracing_attributes.62e245ff-cgu.9.rcgu.o" "/home/build/workspace/aws-sdk/examples/test-utils/target/debug/deps/tracing_attributes-9b799731f627d96b.4n73859jkoxdxj5m.rcgu.rmeta" "/home/build/workspace/aws-sdk/examples/test-utils/target/debug/deps/tracing_attributes-9b799731f627d96b.53km001gtyljpks3.rcgu.o" "-Wl,--as-needed" "-L" "/home/build/workspace/aws-sdk/examples/test-utils/target/debug/deps" "-L" "/opt/rustup/toolchains/1.67.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/home/build/workspace/aws-sdk/examples/test-utils/target/debug/deps/libsyn-5ff6d53c13b63fe1.rlib" "/home/build/workspace/aws-sdk/examples/test-utils/target/debug/deps/libquote-522855fad33e69c5.rlib" "/home/build/workspace/aws-sdk/examples/test-utils/target/debug/deps/libproc_macro2-d0cb96a081d001ab.rlib" "/home/build/workspace/aws-sdk/examples/test-utils/target/debug/deps/libunicode_ident-6f8d5f8ff8a5490f.rlib" "/opt/rustup/toolchains/1.67.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libproc_macro-df9180a12451ef7e.rlib" "/opt/rustup/toolchains/1.67.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-c6192dd4c4d410ac.rlib" "/opt/rustup/toolchains/1.67.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-ff7377081f1949cc.rlib" "/opt/rustup/toolchains/1.67.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libobject-1cb453c5b22d87c0.rlib" "/opt/rustup/toolchains/1.67.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libmemchr-ef80a3f609e31ec2.rlib" "/opt/rustup/toolchains/1.67.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libaddr2line-a690952b7d7394ed.rlib" "/opt/rustup/toolchains/1.67.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgimli-33a38bd462dea42b.rlib" "/opt/rustup/toolchains/1.67.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-2804c2e5e3dadb10.rlib" "/opt/rustup/toolchains/1.67.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_detect-6cea4a991902c072.rlib" "/opt/rustup/toolchains/1.67.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libhashbrown-3f9cac5a3a01ae1f.rlib" "/opt/rustup/toolchains/1.67.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libminiz_oxide-be29f577ccd2e0c7.rlib" "/opt/rustup/toolchains/1.67.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libadler-40cbcc82de628eb3.rlib" "/opt/rustup/toolchains/1.67.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-ce9e735ebca3a756.rlib" "/opt/rustup/toolchains/1.67.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-52730b1d9cce8f6d.rlib" "/opt/rustup/toolchains/1.67.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-dbd9b472d3582c61.rlib" "/opt/rustup/toolchains/1.67.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-8f1d87098c577a9e.rlib" "/opt/rustup/toolchains/1.67.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-72ca4bc0b0597905.rlib" "/opt/rustup/toolchains/1.67.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-cdff4f89b0741001.rlib" "/opt/rustup/toolchains/1.67.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-92115cdd0b50e8be.rlib" "/opt/rustup/toolchains/1.67.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-7f2d6606c0ceff89.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-L" "/opt/rustup/toolchains/1.67.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/home/build/workspace/aws-sdk/examples/test-utils/target/debug/deps/libtracing_attributes-9b799731f627d96b.so" "-Wl,--gc-sections" "-shared" "-Wl,-zrelro,-znow" "-nodefaultlibs"
  = note: collect2: error: ld returned 1 exit status
```

**Comment by thomas-k-cameron** [comment]
I tried to reproduce this but it hasn't been working.

Maybe something to do with the docker image?

**Comment by Velfi** [comment]
> No idea where linker error is coming from.

I've seen this happen as part of the minimal versions check we do. Tracing at v1.0 has some issue with an attribute macro, so when the minimal versions check downgrades tracing, this error can show up. I think we should start using a version of tracing that's newer.

**Comment by jdisanti** [comment]
You don't need to worry about CI on the examples. That's just a notification to us so that we can fix the examples before doing a release. It doesn't block merge. The important CI job that must pass before merge is `ci / Matrix Success`, which did pass here.

File: rust-runtime/aws-smithy-types/src/date_time/de.rs
Comment (jdisanti) [inline rust-runtime/aws-smithy-types/src/date_time/de.rs:34]: Can you make this error message clearer?

**Comment by thomas-k-cameron** [comment]
> You don't need to worry about CI on the examples. That's just a notification to us so that we can fix the examples before doing a release. It doesn't block merge. The important CI job that must pass before merge is `ci / Matrix Success`, which did pass here.

Cool. Thanks!

File: rust-runtime/aws-smithy-types/src/date_time/ser.rs
Comment (jdisanti) [inline rust-runtime/aws-smithy-types/src/date_time/ser.rs:29]: These should be in a `#[cfg(test)] mod tests {`

File: rust-runtime/aws-smithy-types/src/date_time/de.rs
Comment (jdisanti) [inline rust-runtime/aws-smithy-types/src/date_time/de.rs:68]: Same with these

**Review by jdisanti — COMMENTED** [review]
Looking great. Just some minor comments.

File: rust-runtime/aws-smithy-types/src/date_time/ser.rs
Comment (thomas-k-cameron) [reply rust-runtime/aws-smithy-types/src/date_time/ser.rs:28]: fixed.

File: rust-runtime/aws-smithy-types/src/date_time/de.rs
Comment (thomas-k-cameron) [reply rust-runtime/aws-smithy-types/src/date_time/de.rs:76]: fixed

File: rust-runtime/aws-smithy-types/src/date_time/ser.rs
Comment (jdisanti) [inline rust-runtime/aws-smithy-types/src/date_time/ser.rs:28]:
```suggestion
#[cfg(test)]
```

File: rust-runtime/aws-smithy-types/src/date_time/de.rs
Comment (jdisanti) [inline rust-runtime/aws-smithy-types/src/date_time/de.rs:76]:
```suggestion
#[cfg(test)]
```

Comment (thomas-k-cameron) [reply rust-runtime/aws-smithy-types/src/date_time/de.rs]: oops, sorry I pushed before checking. It's fixed now

**Comment by jdisanti** [comment]
I recommend installing the pre-commit commit git hook to the git repo by running `pre-commit install` so that it automatically fixes style issues when committing 😄

**Comment by jdisanti** [comment]
> but it takes for ever to finish (like 10 to 20 minutes).

Huh. That's odd. It takes about 10 seconds on my machine if the tools are already compiled, and maybe 2 minutes otherwise.

**Comment by thomas-k-cameron** [comment]
> > but it takes for ever to finish (like 10 to 20 minutes).
> 
> Huh. That's odd. It takes about 10 seconds on my machine if the tools are already compiled, and maybe 2 minutes otherwise.

Maybe because I'm using VSC on windows inside a docker. no idea.

**Review by jdisanti — APPROVED** [review]

**Pull request merged**
